node/network/collator-protocol/src/collator_side.rs

```diff
@@ -0,0 +1,1009 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+
+use futures::channel::oneshot;
+use log::{trace, warn};
+use polkadot_primitives::v1::{
+	CollatorId, CoreState, Hash, Id as ParaId, CandidateReceipt, PoV, ValidatorId,
+};
+use super::{TARGET,  Result};
+use polkadot_subsystem::{
+	FromOverseer, OverseerSignal, SubsystemContext,
+	messages::{
+		AllMessages, CollatorProtocolMessage, RuntimeApiMessage, RuntimeApiRequest,
+		NetworkBridgeMessage,
+	},
+};
+use polkadot_node_network_protocol::{
+	v1 as protocol_v1, View, PeerId, PeerSet, NetworkBridgeEvent, RequestId,
+};
+use polkadot_node_subsystem_util::{
+	request_validators_ctx,
+	request_validator_groups_ctx,
+};
+
+#[derive(Default)]
+struct State {
+	/// Our id.
+	our_id: CollatorId,
+
+	/// The para this collator is collating on.
+	/// Starts as `None` and is updated with every `CollateOn` message.
+	collating_on: Option<ParaId>,
+
+	/// Track all active peers and their views
+	/// to determine what is relevant to them.
+	peer_views: HashMap<PeerId, View>,
+
+	/// Our own view.
+	view: View,
+
+	/// Posessed collations.
+	///
+	/// We will keep up to one local collation per relay-parent.
+	collations: HashMap<Hash, (CandidateReceipt, PoV)>,
+
+	/// Our validator groups active leafs. 
+	our_validators_groups: HashMap<Hash, Vec<ValidatorId>>,
+
+	/// Validators we know about via `ConnectToValidators` message.
+	/// 
+	/// These are the only validators we are interested in talking to and as such
+	/// all actions from peers not in this map will be ignored.
+	/// Entries in this map will be cleared as validator groups in `our_validator_groups`
+	/// go out of scope with their respective deactivated leafs.
+	known_validators: HashMap<PeerId, ValidatorId>,
+}
+
+/// Distribute a collation.
+/// 
+/// Figure out the core our para is assigned to and the relevant validators.
+/// Issue a connection request to these validators.
+/// If the para is not scheduled or next up on any core, at the relay-parent,
+/// or the relay-parent isn't in the active-leaves set, we ignore the message
+/// as it must be invalid in that case - although this indicates a logic error
+/// elsewhere in the node.
+async fn distribute_collation<Context>(
```

Comment:
This doesn't implement the guide section described [here](https://github.com/paritytech/polkadot/pull/1645/files?file-filters%5B%5D=.md#diff-91227dbbde6d46a6e998ada67f12f84bR114), which is fair, as #1645 is not yet merged. 

It's fine to just say that that's out of scope for this PR and create an issue to handle it later, but we shouldn't forget about it.

Reply:
I think that it does now.